File: frontend/src/features/user/http/api/UpdateUserPermissions.ts
import type { AuthUserListItem } from '@/features/user/model/response/AuthUserListItem'
import type { UpdateUserPermissionsRequest } from '@/features/user/model/request/UpdateUserPermissionsRequest'
import type { Username } from '@/features/user/model/Username'
import { usernameValue } from '@/features/user/lib/user-parsers'
import {
  fromAuthUserListItemContract,
  toUpdateUserPermissionsRequestContract,
} from '@/features/user/http/codec/UserHttpCodecs'
import { postJson } from '@/shared/api/http-client'

export function updateUserPermissions(
  username: Username,
  request: UpdateUserPermissionsRequest,
): Promise<AuthUserListItem> {
  return postJson(
    `/api/users/${encodeURIComponent(usernameValue(username))}/permissions`,
    fromAuthUserListItemContract,
    toUpdateUserPermissionsRequestContract(request),
  )
}

File: backend/src/main/scala/domains/user/http/api/UpdateUserPermissions.scala
package domains.user.http.api



import domains.user.http.*
import domains.user.http.codec.UserHttpCodecs.given
import domains.user.http.mapper.UserHttpRequestMappers
import cats.effect.IO
import domains.user.model.Username
import domains.user.application.UserMutationCommands
import domains.user.http.UserHttpPlanDefinitions.{updateUserPermissions}
import domains.user.model.request.{UpdateUserPermissionsRequest}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.Http4sDsl
import org.http4s.dsl.io.*

object UpdateUserPermissions:

  def routes(context: UserHttpRouteContext)(using Http4sDsl[IO]): HttpRoutes[IO] =
    HttpRoutes.of[IO] {
      case request @ POST -> Root / "api" / "users" / targetUsername / "permissions" =>
        context.handlers.executeDecoded[UpdateUserPermissionsRequest, (Username, UpdateUserPermissionsRequest), UserMutationCommands.UpdateUserPermissionsResult](
          request,
          updateUserPermissions
        )(body => UserHttpRequestMappers.updateUserPermissionsInput(targetUsername, body))
    }
